CBinding/Compiler/ICompiler.cs

```diff
@@ -30,47 +30,431 @@
 // WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 //
 
+using System;
+using System.Threading.Tasks;
+using System.IO;
+using System.Text;
+using System.Linq;
+
 using Mono.Addins;
 
 using MonoDevelop.Core;
 using MonoDevelop.Projects;
+using MonoDevelop.Core.Execution;
+
 
 namespace CBinding
 {
-	[TypeExtensionPoint ("/CBinding/Compilers")]
 	public interface ICompiler
 	{
 		string Name {
 			get;
 		}
-		
+
 		Language Language {
 			get;
 		}
-		
+
 		string CompilerCommand {
 			get;
 		}
-		
+
 		bool SupportsCcache {
 			get;
 		}
-		
+
 		bool SupportsPrecompiledHeaders {
 			get;
 		}
-		
+
 		string GetCompilerFlags (Project project, CProjectConfiguration configuration);
-		
+
 		string GetDefineFlags (Project project, CProjectConfiguration configuration);
-		
+
 		BuildResult Compile (
-		    Project project,
-		    ProjectFileCollection projectFiles,
-		    ProjectPackageCollection packages,
-		    CProjectConfiguration configuration,
-		    ProgressMonitor monitor);
-		
+			Project project,
+			ProjectFileCollection projectFiles,
+			ProjectPackageCollection packages,
+			CProjectConfiguration configuration,
+			ProgressMonitor monitor);
+
 		void Clean (ProjectFileCollection projectFiles, CProjectConfiguration configuration, ProgressMonitor monitor);
 	}
-}
+
+	[TypeExtensionPoint("/CBinding/Toolchains")]
+	public abstract class CMakeToolchain : FolderBasedProject
+	{
+		FilePath file;
+		public override FilePath FileName {
+			get {
+				return file;
+			}
+			set {
+				file = value;
+			}
+		}
+
+		/// The name of this toolchain.
+		public abstract string ToolchainName { 
```

Comment:
Okay, we can still abstract over that :)

Suppose CMakeToolchain has virtual GeneratorID and ToolchainID properties. On the CMakeToolchain class, they'd return null. And GenerateMakefiles treats the -T argument the same way I showed treating the -G argument.

Here's an example:

```csharp

class CMakeToolchain
{
    ///Friendly name for displaying in UI
    public virtual string Name => "Default";

    ///CMake generator ID for this toolchain
    public virtual string GeneratorID => "";

    ///CMake toolchain ID for this toolchain
    public virtual string ToolchainID => "";

    //GenerateMakefiles/ExecuteCommand/GetFileAndLine etc
    //GenerateMakefiles constructs the cmake arguments using GeneratorID and ToolchainID 
}

[Extension("/CBinding/Toolchains")]
class VisualStudio2017LLVMToolchain : CMakeToolchain
{
    public override string Name => "Visual Studio 2017";
    public override string GeneratorID => "Visual Studio 15 2017";
}

[Extension("/CBinding/Toolchains")]
class VisualStudio2017LLVMToolchain : VisualStudio2017Toolchain
{
    public override string Name => "Visual Studio 2017 (LLVM)";
    public override string ToolchainID => "LLVM-vs2014";
}
```

On your example, that